### PR TITLE
Sorted dictionary

### PR DESCRIPTION
**Pull request by lbariogl**
All the elements in the dictionary of cluster-topologies are sorted in decreasing order of frequency, both for common topologies and for groups of rare topologies.

### CODE REVIEW

**Comment by lbariogl** [comment]
Hi @iouribelikov, @shahor02 
this is the version of the software with the completely sorted dictionary.
I attach the final topology distribution.

Sorted_distribution

File: Detectors/ITSMFT/ITS/macros/test/run_buildTopoDict_its.C [hunk 1]
```diff
@@ -202,12 +202,8 @@ void run_buildTopoDict_its(std::string clusfile = "o2clus_its.root",
...
   cComplete->SetLogy();
-  TH1F* hComplete = (TH1F*)dict.mHdist.Clone("hDict");
-  hComplete->SetDirectory(0);
-  hComplete->SetTitle("Topology distribution");
-  hComplete->GetXaxis()->SetTitle("Topology ID");
-  hComplete->SetFillColor(kRed);
-  hComplete->SetFillStyle(3005);
+  TH1F* hComplete = nullptr;
+  o2::itsmft::TopologyDictionary::getTopologyDistribution(dict.getDictionary(), hComplete, "hComplete");
   hComplete->Draw("hist");
```
Comment (shahor02) [inline Detectors/ITSMFT/ITS/macros/test/run_buildTopoDict_its.C:207]: The histo is not shown in the canvas after the macro is run (but is visible in the pdf). The reason is that you create it in the file and it is deleted from memory and canvas when the file is closed. Please add right after its creation:
``hComplete->SetDirctory(0);``.
The same is for other places with similar lines.

Comment (lbariogl) [reply Detectors/ITSMFT/ITS/macros/test/run_buildTopoDict_its.C:208]: Yes, I alway run in batch mode and I did not realise

**Comment by lbariogl** [comment]
Hi @shahor02,
I have implemented all the changes. Just about point 2:

> all macros the std::string inputGeom = "O2geometry.root" need to be changed to std::string inputGeom = "" : recently the naming scheme was changed. I suspect you were working with old simulated data and not up to date dev.

I generated 50 pp collisions using `o2-sim -n 50 -g pythia8 -m PIPE ITS` and `o2-sim-digitizer-workflow` and the file O2geometry.root is generated. Did I use the wrong simulation chain? Or these changes have not been merged yet in dev?

**Comment by shahor02** [comment]
> I generated 50 pp collisions using o2-sim -n 50 -g pythia8 -m PIPE ITS and o2-sim-digitizer-workflow and the file O2geometry.root is generated.

If you see O2geometry.root this means you did not update your dev since ~1week. Or you have this file from old simulation. In the latter case you should see `also o2sim_geometry.root` in the same dir.

Anyway, since there are no conflicts I will merge this PR once it passed the CI tests.

**Comment by iouribelikov** [comment]
@lbariogl Hi Luca.  Indeed, we have now got a few little conflicts. They do not look like difficult to resolve.  In most of the cases, I believe, we may simply retain the fragments from the current dev. 
Let me know if you get any doubts about the resolutions.  Sorry about the inconvenience.
Yura.

**Comment by lbariogl** [comment]
Hi @shahor02, @iouribelikov 
I have solved all the conflicts and made the getters in TopologyDictionay inline as suggested in another pull request

**Comment by shahor02** [comment]
Thanks @lbariogl but the macos CI shows a real error, could you check please?

**Comment by lbariogl** [comment]
Hi @shahor02, @iouribelikov,
actually there are some problems thet were not spotted during my local compilation. Some are trivial, others are less trivial and need discussion. For example, now the macro `CheckTopologies.C` (modified in the last PR), used for the generation of the dictionary, does not work anymore starting from the full cluster ( once, you had to define `_ClusterTopology_` in `Cluster.h`) but only starting from the auxilliary branch with patterns. I do not know if we want to use the additional branch also for the generation of the dictionary. However, I will change the macro `CheckLookUp.C` accordingly to `CheckTopologies.C`, so that everything should compile.

**Comment by shahor02** [comment]
Hi @lbariogl 

The "full clusters" as well as the ``#define _ClusterTopology_`` will be eliminated, the dictionaries should be built from the compact clusters having ``CompCluster::InvalidPatternID`` as a pattern ID.
Could you please verify that it works?

Cheers,
 Ruben

**Comment by iouribelikov** [comment]
@lbariogl @shahor02 Hi Luca, Ruben.  The macro CheckTopologies.C and others have already been modified accordingly in the previous PR.  Just take them fro there. It should be fine.

**Comment by lbariogl** [comment]
Hi @shahor02, @iouribelikov 
after the last changes everything should be ok

**Comment by iouribelikov** [comment]
@lbariogl Great. Thank you for this nice piece of work !

**Comment by shahor02** [comment]
Thanks @lbariogl ! one more point: currently the ``CheckTopologies.C`` will fail on data w/o MC (if hits file is not provided) while it should work also on the raw data. The run_buildTopoDict_its.C is free from this problem (did you check also it, if it works on the new clusters scheme?). I think the protection against absence of the MC should be added to CheckTopologies.C, like in run_buildTopoDict_its.C

**Comment by iouribelikov** [comment]
@shahor02 Yes, the run_buildTopoDict_its.C works with the new cluster scheme.
True, the CheckTopologies.C is calculating the COG offsets with respect to the MC hits.
But, may be, this is OK ?   Assuming "a perfect Monte Carlo", would not we even prefer to have such an option ?

**Comment by iouribelikov** [comment]
@shahor02 Sorry. You we actually suggesting to put a *protection* against missing MC hits.
Yes, this is needed.  Sorry, I missed it while working on my PR.

**Comment by shahor02** [comment]
@iouribelikov yes, normally the dict. will be built using MC to fill pattern properties, but should we be able also to check the patterns of the raw data?

**Comment by lbariogl** [comment]
Hi @shahor02, @iouribelikov 
I have added a protection for MC. Now I have set the difference between MChit and COG (dX and dZ) to zero. So currently the common topologie in dictionary generated with this macro has null error if no MC is used. We could use `dX=BuildTopologyDicionary::IgnoreVal` and in this case use the geometrical information. But maybe this can be addressed in a separate PR.

**Pull request merged**